FIX: correct multichannel couplings

mmikhasenko opened the pull request:
Closes #87

A major problem has been found in implementation of Multichannel BW, in the way how couplings are computed.
The couplings themselves appear only for Flatte. So these values were modified to be consistent with the correct implementation.

- [x] fix json
- [x] fix julia
- [x] fix python 


## 🐛 Bug fixes

- Correct the serialized multichannel Breit-Wigner couplings and deserialize `gsq` as a coupling squared instead of an energy width.

## ✨ New features

- Render a Dalitz plot and the $m_{13}$ and $m_{23}$ mass projections (including per-resonance components) of the deserialized amplitude model.
- Verify the deserialized `generic_function` dynamics (Bugg Breit-Wigner) against the model definition.

## ⚙️ Enhancements

- Render individual dynamical expressions through `unfold_definitions()` so that their complete symbolic structure is visible.
- Assert that each deserialized distribution reproduces the checksums in the model definition, so that the documentation build fails on a mismatch instead of only reporting it in the validation table.
- Document the multichannel Breit-Wigner convention, with references to [`HadronicLineshapes.jl`](https://mmikhasenko.github.io/HadronicLineshapes.jl/dev/10-breitwigner/#Multichannel-Breit-Wigner-Function) and [ComPWA/ampform-dpd#199](https://github.com/ComPWA/ampform-dpd/pull/199).

## 📝 Documentation

- Render figures as SVG with transparent backgrounds and neutral colours, so that they remain readable in dark mode.
- Remove the note block about the amplitude model not being deserialized correctly, which no longer applies.

## 🔨 Maintenance

- Require `ampform-dpd>=0.4.1` and upgrade the lock files accordingly.

## Squash commit messages

```text
* DOC: improve rendering on dark mode
* DOC: remove note block about validation
* ENH: assert equivalence of checksums
* ENH: render sub-expressions with `unfold_definitions()`
* ENH: verify `generic_function`
* FEAT: render Dalitz plot and mass projections
* MAINT: upgrade to ComPWA/ampform-dpd@0.4.1
```

mmikhasenko committed:
fix julia deps

mmikhasenko committed:
fixed json couplings

redeboer commented:
> `MultichannelBreitWigner` is imported, so since it was matched, the problem might be there as well
> 
> ```
> from ampform_dpd.dynamics import (
>     BreitWigner,
>     ChannelArguments,
>     EnergyDependentWidth,
>     MultichannelBreitWigner,
>     P,
>     SimpleBreitWigner,
> )
> ```

These features are not released, see [here](https://github.com/RUB-EP1/amplitude-serialization/blob/32201d80d103379491758a1a67ca93aff38e8908/pixi.toml#L46). But good to be aware for future developments.

mmikhasenko commented:
The problem should be somewhere here, then,

- [definition of MBW](https://github.com/ComPWA/ampform-dpd/blob/5f310882aae1793310cfc4d7ba6b100dc409f180/src/ampform_dpd/dynamics/__init__.py#L207)
- [function to create from dict](https://github.com/ComPWA/ampform-dpd/blob/5f310882aae1793310cfc4d7ba6b100dc409f180/src/ampform_dpd/io/serialization/dynamics.py#L124-L166)

mmikhasenko commented:
@redeboer could you please help understanding your definition,

```
m^2-s-imG
G = sum Gi
#
Gi = 1/m * g^2 * 2p/sqrt(s)
```

In [the code](https://github.com/ComPWA/ampform-dpd/blob/main/src/ampform_dpd/dynamics/__init__.py#L229-L232), I do not recognize the `Gi` expression.

mmikhasenko committed:
replaced julia notebook kernel

mmikhasenko commented:
@redeboer do the tests at the PR check the validation points of json in python code?

I'm surprised to see that the json is modified, python code is not modified, but test button is green.

redeboer commented:
There are some tests, it seems
https://github.com/ComPWA/ampform-dpd/compare/0.2.0...0.2.1rc0#diff-37a837280345b38e5d2b0707adb33df844b8d44971fcae9fcb1aa340073d16cf
But could well be that they don't cover the part of the code that you're referring to.

mmikhasenko commented:
JSON has cross check evaluation, it is surprising that these are not evaluated or pass, given that the lineshape for L1405 changed

redeboer commented:
> it is surprising that these are not evaluated

Not really. Like I said, this was coded together in an afternoon and is therefore only in a release candidate. Serialization is currently not our priority list.

mmikhasenko commented:
This issue is looking forward for being resolved :)

mmikhasenko committed:
Merge branch 'main' into fix-json-multichannelbw

mmikhasenko commented:
ping @redeboer 

Let's check if your `MultichannelBreitWigner` does the same as this,
https://github.com/mmikhasenko/HadronicLineshapes.jl/blob/main/src/shapes.jl#L22-L31

before merging, I would like to see,

1. json is changed, python code gives different results as before (as expected, or as julia)
2. the python code is fixed, dependencies are updated
3. python code gives the same result as expected, or as julia

redeboer committed:
Merge remote-tracking branch 'origin/main' into fix-json-multichannelbw

redeboer committed:
DX: activate Jupyter extension for interactive Quarto

redeboer committed:
MAINT: upgrade to AmpForm-DPD v0.3.0

redeboer committed:
ENH: do not set `indent_size` for `qmd` files

redeboer committed:
DX: activate Ruff formatting in VS Code

redeboer committed:
ENH: improve validation point rendering

redeboer committed:
MAINT: upgrade lock files

redeboer committed:
Merge branch 'main' into fix-json-multichannelbw

redeboer committed:
Merge branch 'main' into fix-json-multichannelbw

redeboer committed:
FIX: hash caches with regard to `qmd` files

redeboer committed:
FIX: add JSON models to cach hashes as well

redeboer committed:
DX: implement validation table

redeboer committed:
DX: define `pixi run style` task

redeboer committed:
DX: define `pixi run all` task

redeboer committed:
TEST: install ComPWA/ampform-dpd@ef80b1a

redeboer committed:
DX: push preview for GitHub pages

redeboer committed:
Merge remote-tracking branch 'origin/main' into fix-json-multichannelbw

redeboer committed:
FIX: remove Jupyter caches in `doc` task

redeboer committed:
DOC: remove note block about validation

redeboer committed:
DOC: add Demos logo and favicon

redeboer committed:
FIX: remove `language_version` from `cspell` hook

redeboer committed:
DOC: remove redundant definitions

redeboer committed:
MAINT: upgrade to ComPWA/ampform-dpd@6ba6a85

redeboer committed:
ENH: render sub-expressions with `unfold_definitions()`

dependabot[bot] committed:
MAINT: Bump https://github.com/tombi-toml/tombi-pre-commit

Bumps the lock group with 1 update: [https://github.com/tombi-toml/tombi-pre-commit](https://github.com/tombi-toml/tombi-pre-commit).


Updates `https://github.com/tombi-toml/tombi-pre-commit` from v1.2.0 to 1.2.5
- [Release notes](https://github.com/tombi-toml/tombi-pre-commit/releases)
- [Commits](https://github.com/tombi-toml/tombi-pre-commit/compare/v1.2.0...v1.2.5)

---
updated-dependencies:
- dependency-name: https://github.com/tombi-toml/tombi-pre-commit
  dependency-version: 1.2.5
  dependency-type: direct:production
  dependency-group: lock
...

Signed-off-by: dependabot[bot] <support@github.com>

redeboer committed:
MAINT: upgrade lock files

redeboer committed:
DOC: update favicon and logo

redeboer committed:
DX: remove Jupyter caches before building docs

redeboer committed:
MAINT: stack PR #90 on PR #100

redeboer committed:
ENH: verify `generic_function`

redeboer committed:
MAINT: upgrade lock files

redeboer committed:
MAINT: install ComPWA/ampform-dpd@fbcca49

redeboer committed:
MAINT: upgrade lock files

redeboer committed:
Merge remote-tracking branch 'origin/dependabot/lock-ee0317cf3e' into fix-json-multichannelbw

redeboer committed:
MAINT: upgrade to ComPWA/ampform-dpd@e68a987

redeboer committed:
FEAT: assert equivalence of checksums

redeboer committed:
DOC: improve rendering on dark mode

redeboer committed:
DX: update `docs/.gitignore`

redeboer committed:
DX: update `docs/.gitignore`

redeboer committed:
DOC: render Dalitz plot and mass projections

redeboer committed:
MAINT: minor formatting improvements

redeboer committed:
MAINT: upgrade to ComPWA/ampform-dpd@0.4.1

redeboer committed:
Merge branch 'dependabot/lock-ee0317cf3e' into fix-json-multichannelbw

redeboer committed:
Merge branch 'main' into fix-json-multichannelbw
